File: graph/arch/1584-min-cost-connect-points/union-find.js
class UnionFind {
  constructor(n) {
    this.rs = Array.from(Array(n), (_, i) => i);
    this.hs = Array(n).fill(0);
  }
  find(i) { 
    let r = this.rs[i]; 
    while (i != r) { i = r; r = this.rs[i] }
    return r;
  }
  connect(i, j) { 
    const ri = this.find(i), rj = this.find(j);
    if (ri == rj) { return false }
    const hi = this.hs[ri], hj = this.hs[rj];
    if (hi < hj) { this.rs[ri] = rj }
    else if (hj < hi) { this.rs[rj] = ri }
    else { this.rs[ri] = rj; ++this.hs[rj] }
    return true;
  }
}

module.exports = UnionFind;

File: graph/arch/1584-min-cost-connect-points/py/union_find.py
class UnionFind:
  def __init__(self, n: int):
    self.rs = [i for i in range(n)]
    self.hs = [1]*n

  def find(self, i: int) -> int:
    while i != self.rs[i]:
      i = self.rs[i]
    return i

  def connect(self, i: int, j: int) -> bool:
    ri, rj = self.find(i), self.find(j)
    if ri == rj:
      return False
    if self.hs[ri] < self.hs[rj]:
      self.rs[ri] = rj
    elif self.hs[rj] < self.hs[ri]:
      self.rs[rj] = ri
    else:
      self.rs[ri] = rj
      self.hs[rj] = self.hs[rj]+1
    return True
